Implement GH-15711: Allow SoapClient to use the backing value during response serialization

nielsdos opened the pull request:
See individual commits, some code had to be adapted first.

Girgias commented on ext/soap/php_encoding.c:
```diff
@@ -830,33 +831,34 @@ static zval *to_zval_hexbin(zval *ret, encodeTypePtr type, xmlNodePtr data)
 	return ret;
 }
 
+static zend_string *get_serialization_string_from_zval(zval *data)
```
Should this have a "SOAP" prefix? I don't remember if SOAP uses them or  not

nielsdos replied:
The other methods here don't do that, most methods in soap actually don't :shrug:

Girgias reviewed (approved):
The overall logic makes sense to me, possibly should be backported as it could be considered a bug?

nielsdos commented:
I updated one commit to remove a dead break statement, and I added a commit adding support for int-backed enums as per https://github.com/php/php-src/issues/15711#issuecomment-2337397874

nielsdos commented:
> Possibly should be backported as it could be considered a bug?

I don't know, I think this is closer to a feature than a bugfix.

nielsdos commented:
Had to rebase to make CI green again due to master failure

Girgias commented on ext/soap/php_encoding.c:
```diff
@@ -850,6 +850,30 @@ static zend_string *get_serialization_string_from_zval(zval *data)
+			}
+			ZEND_FALLTHROUGH;
+		default:
+			return zval_get_long(data);
```
Is this guaranteed to be an integer string?

nielsdos replied:
No. It'll throw if that's not the case.

MrMeshok commented:
> > Possibly should be backported as it could be considered a bug?
> 
> I don't know, I think this is closer to a feature than a bugfix.

I would appreciate backport to 8.3. But if not, it's not a big deal. I'll just add TODOs to remove workarounds when 8.4 releases

nielsdos commented:
Afaik today is the last day we can still add features to 8.4. The question is whether this is a feature or a bugfix. Cc @SakiTakamachi @ericmann @NattyNarwhal

SakiTakamachi commented:
@nielsdos 
I would consider this mostly a bug fix.

@NattyNarwhal 
How do you think?

NattyNarwhal commented:
Unfortunately I just cut the release already. It might not be too late to get a feature in though, from the sounds of [this comment](https://github.com/php/php-src/issues/15796#issuecomment-2338522161), it might be possible to get it after beta but before RC, but I could be misreading it though.

That said, I am sympathetic to call this a bug fix based on the [original issue's description](https://github.com/php/php-src/issues/15711#issuecomment-2336595751), since it's an inconsistency between places where you'd want to serialize values.

nielsdos commented:
I'll wait for Eric to see if he sees this too as a bugfix, and in that case there's no issue at all.

> Unfortunately I just cut the release already. It might not be too late to get a feature in though, from the sounds of [this comment](https://github.com/php/php-src/issues/15796#issuecomment-2338522161), it might be possible to get it after beta but before RC, but I could be misreading it though.

I am reading https://wiki.php.net/rfc/release_cycle_update and getting confused:

One the one hand it states:

> We propose to drop this small feature exception, and disallow any feature addition in Release Candidates, Bug Fix releases, and of course, Security releases. 

But it also states (under "Revised timeline"):

> Sep 26 2024: RC 1 — No more new features may be introduced into the 8.4 release from now on.

So this seems contradictory?

ericmann commented:
> I'll wait for Eric to see if he sees this too as a bugfix, and in that case there's no issue at all.

I don't have a strong feeling either way, but this does _feel_ like a bugfix. I'd like to see it land in both 8.3 _and_ 8.4 though ...

nielsdos commented:
Thanks for the feedback, I'll reclassify the issue report and get this into 8.3 & master.
